Error message shows type of private field

The code below yields two errors:

1. The field is private
2. Type error in if expression because type is `<type of private field>` instead of `bool`

This means a user can observe the type of a private field of a dependency in their error messages. It doesn't mean they can actually use that information, but it seems weird at least.

Is this actually an issue? If you don't consider it one, feel free to close this.

- - -

```rust
fn main() {
    let x = foo::Foo::default();
    if x.len {
        println!("foo");
    }
}

mod foo {
    #[derive(Default)]
    pub struct Foo {
        len: String,
    }
    
    impl Foo {
        pub fn len(&self) -> usize {
            42
        }
    }
}
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5bca9cfb20ef8a1a4d637932a742510b)

(originally seen [on reddit](https://www.reddit.com/r/rust/comments/acapjb/rust_error_messages_are_so_quite_3/))

I won't close the ticket as this deserves a conversation to be had, but I _personally_ don't consider it to 
be a big problem.

---

Parsing x.len as `x.len()` when we suggest that would be nice, but in this case it would lead to confusing errors if we don't annotate the compiler state with something along the lines of

```
error[E0616]: field `len` of struct `foo::Foo` is private
 --> src/main.rs:3:8
  |
3 |     if x.len {
  |        ^^^^^ help: a method `len` also exists, perhaps you wish to call it: `x.len()`

error[E0308]: mismatched types
 --> src/main.rs:3:8
  |
3 |     if x.len {
  |        ^^^^^ expected bool, found `usize`
  |
  = note: expected type `bool`
             found type `usize`
  = note: found `usize` assuming that `x.len` has been rewritten to `x.len()`
```

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -2217,7 +2217,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             self.tcx.check_stability(field.did, Some(expr.hir_id), expr.span, None);
                             return field_ty;
                         }
-                        private_candidate = Some((adjustments, base_def.did(), field_ty));
+                        private_candidate = Some((adjustments, base_def.did()));
                     }
                 }
                 ty::Tuple(tys) => {
@@ -2240,12 +2240,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
         self.structurally_resolved_type(autoderef.span(), autoderef.final_ty(false));
 
-        if let Some((adjustments, did, field_ty)) = private_candidate {
+        if let Some((adjustments, did)) = private_candidate {
             // (#90483) apply adjustments to avoid ExprUseVisitor from
             // creating erroneous projection.
             self.apply_adjustments(base, adjustments);
             self.ban_private_field_access(expr, base_ty, field, did);
-            return field_ty;
+            return self.tcx().ty_error();
         }
 
         if field.name == kw::Empty {

--- a/src/test/ui/issues/issue-25386.rs
+++ b/src/test/ui/issues/issue-25386.rs
@@ -24,5 +24,4 @@ macro_rules! check_ptr_exist {
 fn main() {
     let item = stuff::Item::new();
     println!("{}", check_ptr_exist!(item, name));
-    //~^ ERROR field `name` of struct `CObj` is private
 }

--- a/src/test/ui/issues/issue-25386.stderr
+++ b/src/test/ui/issues/issue-25386.stderr
@@ -9,12 +9,6 @@ LL |     println!("{}", check_ptr_exist!(item, name));
    |
    = note: this error originates in the macro `check_ptr_exist` (in Nightly builds, run with -Z macro-backtrace for more info)
 
-error[E0616]: field `name` of struct `CObj` is private
-  --> $DIR/issue-25386.rs:26:43
-   |
-LL |     println!("{}", check_ptr_exist!(item, name));
-   |                                           ^^^^ private field
-
-error: aborting due to 2 previous errors
+error: aborting due to previous error
 
 For more information about this error, try `rustc --explain E0616`.

--- a/src/test/ui/privacy/private-field-ty-err.rs
+++ b/src/test/ui/privacy/private-field-ty-err.rs
@@ -1,0 +1,20 @@
+fn main() {
+    let x = foo::Foo::default();
+    if x.len {
+        //~^ ERROR field `len` of struct `Foo` is private
+        println!("foo");
+    }
+}
+
+mod foo {
+    #[derive(Default)]
+    pub struct Foo {
+        len: String,
+    }
+
+    impl Foo {
+        pub fn len(&self) -> usize {
+            42
+        }
+    }
+}

--- a/src/test/ui/privacy/private-field-ty-err.stderr
+++ b/src/test/ui/privacy/private-field-ty-err.stderr
@@ -1,0 +1,14 @@
+error[E0616]: field `len` of struct `Foo` is private
+  --> $DIR/private-field-ty-err.rs:3:10
+   |
+LL |     if x.len {
+   |          ^^^ private field
+   |
+help: a method `len` also exists, call it with parentheses
+   |
+LL |     if x.len() {
+   |             ++
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0616`.

